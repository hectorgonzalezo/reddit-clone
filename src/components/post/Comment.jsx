import React, { useState, useEffect } from 'react';
import {
  arrayOf,
  objectOf,
  oneOfType,
  string,
  number,
  array,
} from 'prop-types';
import { formatDistanceToNow } from 'date-fns';
import { database } from '../../firebase/firebase';

function Comment({ comment }) {
  const [userIcon, setUserIcon] = useState('');

  // get user icon from database
  useEffect(() => {
    database.getUser(comment.user).then((user) => setUserIcon(user.icon));
  }, []);

  return (
    <div className="comment">
      <div className="poster-area">
        {userIcon !== '' ? <img src={userIcon} className="user-icon" alt="user icon" /> : null}
        <a href="">{comment.user}</a>
        <p>
          &nbsp;â€¢&nbsp;
          {formatDistanceToNow(new Date(comment.timePosted))}
          &nbsp;ago
        </p>
      </div>
      <div>
        <p>{comment.content}</p>
      </div>
    </div>
  );
}

Comment.propTypes = {
  comment: objectOf(oneOfType([string, number, array])).isRequired,
};

export default Comment;
