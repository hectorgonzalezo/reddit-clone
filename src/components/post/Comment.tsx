import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import { formatDistanceToNow } from 'date-fns';
import { selectUser } from '../../store/userSlice';
import CommentCreator from './CommentCreator';
import commentIcon from '../../assets/comments_icon.svg';
import { getUser } from '../../api/users';
import defaultUserIcon from '../../assets/user_icon.svg';

interface CommentProps {
  commentsList: IComment[];
  commentIndex: number[];
  comment: IComment;
  subreddit: string;
  postId: string;
  reloadPost: () => void;
};


function Comment({
  comment,
  commentsList = [],
  commentIndex = [0],
  subreddit,
  postId,
  reloadPost,
}: CommentProps): JSX.Element {
  const user = useSelector(selectUser);
  const [userIcon, setUserIcon] = useState(defaultUserIcon);
  const [visibleCreator, setVisibleCreator] = useState(false);
  const [visibleEditor, setVisibleEditor] = useState(false);
  // commentIndex contains the index to find the comment in the nested
  // hierarchy of comments in the database

  // get user icon from database
  useEffect(() => {
    getUser(comment.user.toString()).then((fetchedUser) => {
      // if theres an icon, add it to state
      if (fetchedUser?.icon !== undefined) {
        // console.log()
        setUserIcon(fetchedUser.icon);
      }
    })
    .catch((error) => console.log(error));
  }, []);

  return (
    <div className="comment-inside" data-testid="comment">
      <div className="comment">
        <div className="poster-area">
          <img src={userIcon} className="user-icon" alt="user icon" />
          <a href="">{comment.user}</a>
          <p>
            &nbsp;â€¢&nbsp;
            {formatDistanceToNow(new Date(comment.createdAt))}
            &nbsp;ago
          </p>
        </div>
        {visibleEditor ? (
          <CommentCreator
            commentsList={commentsList}
            value={comment.text}
            index={commentIndex}
            subreddit={subreddit}
            postId={postId}
            reloadPost={reloadPost}
            edit
          />
        ) : (
          <p>{comment.text}</p>
        )}
        <div>
          <button
            type="button"
            onClick={() => setVisibleCreator((prev) => !prev)}
            data-testid="reply-button"
          >
            <img src={commentIcon} alt="" className="icon" />
            <p>Reply</p>
          </button>
          {comment.user === user._id.toString() ? (
            <button
              type="button"
              onClick={() => setVisibleEditor((prev) => !prev)}
              data-testid="edit-button"
            >
              <p>Edit</p>
            </button>
          ) : null}
        </div>
        {visibleCreator ? (
          <CommentCreator
            commentsList={commentsList}
            index={commentIndex}
            subreddit={subreddit}
            postId={postId}
            reloadPost={reloadPost}
          />
        ) : null}
      </div>
      {/* show responses if there are any */}
      {comment.responses !== undefined
        ? comment.responses.map((response, newIndex) => (
          <Comment
            key={newIndex}
            commentsList={commentsList}
            commentIndex={commentIndex.concat(newIndex)}
            comment={response}
            subreddit={subreddit}
            postId={postId}
            reloadPost={reloadPost}
          />
        ))
        : null}
    </div>
  );
}



export default Comment;
