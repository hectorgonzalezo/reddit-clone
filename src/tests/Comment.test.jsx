import React from 'react';
import { render, screen, container, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import Comment from '../components/post/Comment';
jest.mock('..//firebase/firebase');

describe('Comment', () => {
  test('Display basic comment data', async () => {
    const mockComment = {
      timePosted: (new Date()).toString(),
      content: 'mockContent',
      user: 'mockUser',
      upVotes: 10,
    };

    await act(async () => render(<Comment comment={mockComment} />));

    expect(screen.getByText('mockContent')).toBeInTheDocument();
    expect(screen.getByText('mockUser')).toBeInTheDocument();
    expect(screen.getByText(/less/)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
