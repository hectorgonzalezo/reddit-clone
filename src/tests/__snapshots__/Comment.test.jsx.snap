// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Display basic comment data 1`] = `undefined`;
